projects/util-linux-2.40.2/misc-utils/pipesz.1
'\" t
.\"     Title: pipesz
.\"    Author: [see the "AUTHOR(S)" section]
.\" Generator: Asciidoctor 2.0.20
.\"      Date: 2024-04-24
.\"    Manual: User Commands
.\"    Source: util-linux 2.40.2
.\"  Language: English
.\"
.TH "PIPESZ" "1" "2024-04-24" "util\-linux 2.40.2" "User Commands"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\fI\\$2\fP <\\$1>\\$3
..
.als MTO URL
.if \n[.g] \{\
.  mso www.tmac
.  am URL
.    ad l
.  .
.  am MTO
.    ad l
.  .
.  LINKSTYLE blue R < >
.\}
.SH "NAME"
pipesz \- set or examine pipe and FIFO buffer sizes
.SH "SYNOPSIS"
.sp
\fBpipesz\fP [options] [\-\-set \fIsize\fP] [\-\-] [\fIcommand\fP [argument] ...]
.sp
\fBpipesz\fP [options] \-\-get
.SH "DESCRIPTION"
.sp
Pipes and FIFOs maintain an internal buffer used to transfer data between the read end and the write end. In some cases, the default size of this internal buffer may not be appropriate. This program provides facilities to set and examine the size of these buffers.
.sp
The \fB\-\-set\fP operation sets pipe buffer sizes. If it is specified, it must be specified with an explicit \fIsize\fP. Otherwise, it is implied and the size is read from \fB/proc/sys/fs/pipe\-max\-size\fP. The kernel may adjust \fIsize\fP as described in \fBfcntl\fP(2). To determine the actual buffer sizes set, use the \fB\-\-verbose\fP option. If neither \fB\-\-file\fP nor \fB\-\-fd\fP are specified, \fB\-\-set\fP acts on standard output.
.sp
The \fB\-\-set\fP operation permits an optional \fIcommand\fP to execute after setting the pipe buffer sizes. This command is executed with the adjusted pipes.
.sp
The \fB\-\-get\fP operation outputs data in a tabular format. The first column is the name of the pipe as passed to \fBpipesz\fP. File descriptors are named as "fd \fIN\fP". The second column is the size, in bytes, of the pipe\(cqs internal buffer. The third column is the number of unread bytes currently in the pipe. The columns are separated by tabs (\*(Aq\(rst\*(Aq, ASCII 09h). If \fB\-\-verbose\fP is specified, a descriptive header is also emitted. If neither \fB\-\-file\fP nor \fB\-\-fd\fP are specified, \fB\-\-get\fP acts on standard input.
.sp
Unless the \fB\-\-check\fP option is specified, \fBpipesz\fP does \fInot\fP exit if it encounters an error while manipulating a file or file descriptor. This allows \fBpipesz\fP to be used generically without fear of disrupting the execution of pipelines should the type of certain files be later changed. For minimal disruption, the \fB\-\-quiet\fP option prevents warnings from being emitted in these cases.
.sp
The kernel imposes limits on the amount of pipe buffer space unprivileged processes can use, though see \fBBUGS\fP below. The kernel will also refuse to shrink a pipe buffer if this would cause a loss of buffered data. See \fBpipe\fP(7) for additional details.
.sp
\fBpipesz\fP supports specifying multiple short options consecutively, in the usual \fBgetopt\fP(3) fashion. The first non\-option argument is interpreted as \fIcommand\fP. If \fIcommand\fP might begin with \*(Aq\-\*(Aq, use \*(Aq\-\-\*(Aq to separate it from arguments to \fBpipesz\fP. In shell scripts, it is good practice to use \*(Aq\-\-\*(Aq when parameter expansion is involved. \fBpipesz\fP itself does not read from standard input and does not write to standard output unless \fB\-\-get\fP, \fB\-\-help\fP, or \fB\-\-version\fP are specified.
.SH "OPTIONS"
.sp
\fB\-g\fP, \fB\-\-get\fP
.RS 4
Report the size of pipe buffers to standard output and exit. As a special behavior, if neither \fB\-\-file\fP nor \fB\-\-fd\fP are specified, standard input is examined. It is an error to specify this option in combination with \fB\-\-set\fP.
.RE
.sp
\fB\-s\fP, \fB\-\-set\fP \fIsize\fP
.RS 4
Set the size of the pipe buffers, in bytes. This option may be suffixed by \fIK\fP, \fIM\fP, \fIG\fP, \fIKiB\fP, \fIMiB\fP, or \fIGiB\fP to indicate multiples of 1024. Fractional values are supported in this case. Additional suffixes are supported but are unlikely to be useful. If this option is not specified, a default value is used, as described above. If this option is specified multiple times, a warning is emitted and only the last\-specified \fIsize\fP is used. As a special behavior, if neither \fB\-\-file\fP nor \fB\-\-fd\fP are specified, standard output is adjusted. It is an error to specify this option in combination with \fB\-\-get\fP.
.RE
.sp
\fB\-f\fP, \fB\-\-file\fP \fIpath\fP
.RS 4
Set the buffer size of the FIFO or pipe at \fIpath\fP, relative to the current working directory. You may specify this option multiple times to affect different files, and you may do so in combination with \fB\-\-fd\fP. Generally, this option is used with FIFOs, but it will also operate on anonymous pipes such as those found in \fB/proc/PID/fd\fP. Changes to the buffer size of FIFOs are not preserved across system restarts.
.RE
.sp
\fB\-n\fP, \fB\-\-fd\fP \fIfd\fP
.RS 4
Set the buffer size of the pipe or FIFO passed to \fBpipesz\fP as the specified file descriptor number. You may specify this option multiple times to affect different file descriptors, and you may do so in combination with \fB\-\-file\fP. Shorthand options are provided for the common cases of fd 0 (stdin), fd 1 (stdout), and fd 2 (stderr). These should suffice in most cases.
.RE
.sp
\fB\-i\fP, \fB\-\-stdin\fP
.RS 4
Shorthand for \fB\-\-fd 0\fP.
.RE
.sp
\fB\-o\fP, \fB\-\-stdout\fP
.RS 4
Shorthand for \fB\-\-fd 1\fP.
.RE
.sp
\fB\-e\fP, \fB\-\-stderr\fP
.RS 4
Shorthand for \fB\-\-fd 2\fP.
.RE
.sp
\fB\-c\fP, \fB\-\-check\fP
.RS 4
Exit, without executing \fIcommand\fP, in case of any error while manipulating a file or file descriptor. The default behavior if this is not specified is to emit a warning to standard error and continue.
.RE
.sp
\fB\-q\fP, \fB\-\-quiet\fP
.RS 4
Do not diagnose non\-fatal errors to standard error. This option does not affect the normal output of \fB\-\-get\fP, \fB\-\-verbose\fP, \fB\-\-help\fP, or \fB\-\-version\fP.
.RE
.sp
\fB\-v\fP, \fB\-\-verbose\fP
.RS 4
If specified with \fB\-\-get\fP, \fBpipesz\fP will emit a descriptive header above the table. Otherwise, if specified, \fBpipesz\fP will print the actual buffer sizes set by the kernel to standard error.
.RE
.sp
\fB\-h\fP, \fB\-\-help\fP
.RS 4
Display help text and exit.
.RE
.sp
\fB\-V\fP, \fB\-\-version\fP
.RS 4
Print version and exit.
.RE
.SH "EXAMPLES"
.sp
\fBpipesz\fP \fBdd\fP if=\fIfile\fP bs=1M | ...
.RS 4
Runs \fBdd\fP(1) with an expanded standard output pipe, allowing it to avoid context switches when piping around large blocks.
.RE
.sp
\fBpipesz\fP \-s1M \-cf \fI/run/my\-service.fifo\fP
.RS 4
Sets the pipe buffer size of a service FIFO to 1,048,576 bytes. If the buffer size could not be set, \fBpipesz\fP exits with an error.
.RE
.sp
\fBecho\fP hello | \fBpipesz\fP \-g
.RS 4
Prints the size of pipe used by the shell to pass input to \fBpipesz\fP. Since \fBpipesz\fP does not read standard input, it may also report 6 unread bytes in the pipe, depending on relative timings.
.RE
.sp
\fBfind\fP /proc/\fIPID\fP/fd \-exec \fBpipesz\fP \-gqf \*(Aq{}\*(Aq \*(Aq;\*(Aq
.RS 4
Prints the size and number of unread bytes of all pipes in use by \fIPID\fP. If some pipes are routinely full, \fBpipesz\fP might be able to mitigate a processing bottleneck.
.RE
.SH "NOTES"
.sp
Linux supports adjusting the size of pipe buffers since kernel 2.6.35. This release also introduced \fB/proc/sys/fs/pipe\-max\-size\fP.
.sp
This program uses \fBfcntl\fP(2) \fBF_GETPIPE_SZ\fP/\fBF_SETPIPE_SZ\fP to get and set pipe buffer sizes.
.sp
This program uses \fBioctl\fP(2) \fBFIONREAD\fP to report the amount of unread data in pipes. If for some reason this fails, the amount of unread data is reported as 0.
.SH "BUGS"
.sp
Before Linux 4.9, some bugs affect how certain resource limits are enforced when setting pipe buffer sizes. See \fBpipe\fP(7) for details.
.SH "AUTHORS"
.sp
.MTO "nwsharp\(atlive.com" "Nathan Sharp" ""
.SH "SEE ALSO"
.sp
\fBpipe\fP(7)
.SH "REPORTING BUGS"
.sp
For bug reports, use the issue tracker at \c
.URL "https://github.com/util\-linux/util\-linux/issues" "" "."
.SH "AVAILABILITY"
.sp
The \fBpipesz\fP command is part of the util\-linux package which can be downloaded from \c
.URL "https://www.kernel.org/pub/linux/utils/util\-linux/" "Linux Kernel Archive" "."

projects/util-linux-2.40.2/misc-utils/pipesz.c
/*
 * pipesz(1) - Set or examine pipe buffer sizes.
 *
 * Copyright (c) 2022 Nathan Sharp
 * Written by Nathan Sharp <nwsharp@live.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it would be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <getopt.h>
#include <sys/ioctl.h>		/* FIONREAD */
#include <fcntl.h>		/* F_GETPIPE_SZ F_SETPIPE_SZ */

#include "c.h"
#include "nls.h"

#include "closestream.h"	/* close_stdout_atexit */
#include "optutils.h"		/* err_exclusive_options */
#include "path.h"		/* ul_path_read_s32 */
#include "pathnames.h"		/* _PATH_PROC_PIPE_MAX_SIZE */
#include "strutils.h"		/* strtos32_or_err strtosize_or_err */

static char opt_check = 0;	/* --check */
static char opt_get = 0;	/* --get */
static char opt_quiet = 0;	/* --quiet */
static int opt_size = -1;	/* --set <size> */
static char opt_verbose = 0;	/* --verbose */

/* fallback file for default size */
#ifndef PIPESZ_DEFAULT_SIZE_FILE
#define PIPESZ_DEFAULT_SIZE_FILE _PATH_PROC_PIPE_MAX_SIZE
#endif

/* convenience macros, since pipesz is by default very lenient */
#define check(FMT...) do {			\
	if (opt_check) {			\
		err(EXIT_FAILURE, FMT);		\
	} else if (!opt_quiet)	{		\
		warn(FMT);			\
	}					\
} while (0)

#define checkx(FMT...) do {			\
	if (opt_check) {			\
		errx(EXIT_FAILURE, FMT);	\
	} else if (!opt_quiet) {		\
		warnx(FMT);			\
	}					\
} while (0)

static void __attribute__((__noreturn__)) usage(void)
{
	fputs(USAGE_HEADER, stdout);
	fprintf(stdout, _(" %s [options] [--set <size>] [--] [command]\n"), program_invocation_short_name);
	fprintf(stdout, _(" %s [options] --get\n"), program_invocation_short_name);

	fputs(USAGE_SEPARATOR, stdout);
	/* TRANSLATORS: 'command' refers to a program argument */
	fputsln(_("Set or examine pipe buffer sizes and optionally execute command."), stdout);

	fputs(USAGE_OPTIONS, stdout);
	fputsln(_(" -g, --get          examine pipe buffers"), stdout);
	/* TRANSLATORS: '%s' refers to a system file */
	fprintf(stdout,
	     _(" -s, --set <size>   set pipe buffer sizes\n"
	       "                      size defaults to %s\n"),
		PIPESZ_DEFAULT_SIZE_FILE);

	fputs(USAGE_SEPARATOR, stdout);
	fputsln(_(" -f, --file <path>  act on a file"), stdout);
	fputsln(_(" -n, --fd <num>     act on a file descriptor"), stdout);
	fputsln(_(" -i, --stdin        act on standard input"), stdout);
	fputsln(_(" -o, --stdout       act on standard output"), stdout);
	fputsln(_(" -e, --stderr       act on standard error"), stdout);

	fputs(USAGE_SEPARATOR, stdout);
	fputsln(_(" -c, --check        do not continue after an error"), stdout);
	fputsln(_(" -q, --quiet        do not warn of non-fatal errors"), stdout);
	fputsln(_(" -v, --verbose      provide detailed output"), stdout);

	fputs(USAGE_SEPARATOR, stdout);
	fprintf(stdout, USAGE_HELP_OPTIONS(20));

	fprintf(stdout, USAGE_MAN_TAIL("pipesz(1)"));

	exit(EXIT_SUCCESS);
}

/*
 * performs F_GETPIPE_SZ and FIONREAD
 * outputs a table row
 */
static void do_get(int fd, const char *name)
{
	int sz, used;

	sz = fcntl(fd, F_GETPIPE_SZ);
	if (sz < 0) {
		/* TRANSLATORS: '%s' refers to a file */
		check(_("cannot get pipe buffer size of %s"), name);
		return;
	}

	if (ioctl(fd, FIONREAD, &used))
		used = 0;

	printf("%s\t%d\t%d\n", name, sz, used);
}

/*
 * performs F_SETPIPE_SZ
 */
static void do_set(int fd, const char *name)
{
	int sz;

	sz = fcntl(fd, F_SETPIPE_SZ, opt_size);
	if (sz < 0)
		/* TRANSLATORS: '%s' refers to a file */
		check(_("cannot set pipe buffer size of %s"), name);
	else if (opt_verbose)
		/* TRANSLATORS: '%s' refers to a file, '%d' to a buffer size in bytes */
		warnx(_("%s pipe buffer size set to %d"), name, sz);
}

/*
 * does the requested operation on an fd
 */
static void do_fd(int fd)
{
	char name[sizeof(stringify(INT_MIN)) + 3];

	sprintf(name, "fd %d", fd);

	if (opt_get)
		do_get(fd, name);
	else
		do_set(fd, name);
}

/*
 * does the requested operation on a file
 */
static void do_file(const char *path)
{
	int fd;

	fd = open(path, O_RDONLY | O_CLOEXEC);
	if (fd < 0) {
		/* TRANSLATORS: '%s' refers to a file */
		check(_("cannot open %s"), path);
		return;
	}

	if (opt_get)
		do_get(fd, path);
	else
		do_set(fd, path);

	close(fd);
}

/*
 * if necessary, determines a default buffer size and places it in opt_size
 * returns FALSE if this could not be done
 */
static char set_size_default(void)
{
	if (opt_size >= 0)
		return TRUE;

	if (ul_path_read_s32(NULL, &opt_size, PIPESZ_DEFAULT_SIZE_FILE)) {
		/* TRANSLATORS: '%s' refers to a system file */
		check(_("cannot parse %s"), PIPESZ_DEFAULT_SIZE_FILE);
		return FALSE;
	}

	if (opt_size < 0) {
		/* TRANSLATORS: '%s' refers to a system file */
		checkx(_("cannot parse %s"), PIPESZ_DEFAULT_SIZE_FILE);
		return FALSE;
	}

	return TRUE;
}

int main(int argc, char **argv)
{
	static const char shortopts[] = "+cef:ghin:oqs:vV";
	static const struct option longopts[] = {
		{ "check",     no_argument,       NULL, 'c' },
		{ "fd",        required_argument, NULL, 'n' },
		{ "file",      required_argument, NULL, 'f' },
		{ "get",       no_argument,       NULL, 'g' },
		{ "help",      no_argument,       NULL, 'h' },
		{ "quiet",     no_argument,       NULL, 'q' },
		{ "set",       required_argument, NULL, 's' },
		{ "stdin",     no_argument,       NULL, 'i' },
		{ "stdout",    no_argument,       NULL, 'o' },
		{ "stderr",    no_argument,       NULL, 'e' },
		{ "verbose",   no_argument,       NULL, 'v' },
		{ "version",   no_argument,       NULL, 'V' },
		{ NULL, 0, NULL, 0 }
	};
	static const ul_excl_t excl[] = {
		{ 'g', 's' },
		{ 0 }
	};

	int c, fd, n_opt_pipe = 0, n_opt_size = 0;
	uintmax_t sz;

	int excl_st[ARRAY_SIZE(excl)] = UL_EXCL_STATUS_INIT;

	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
	close_stdout_atexit();

	/* check for --help or --version */
	while ((c = getopt_long(argc, argv, shortopts, longopts, NULL)) != -1) {
		switch (c) {
		case 'h':
			usage();
		case 'V':
			print_version(EXIT_SUCCESS);
		}
	}

	/* gather normal options */
	optind = 1;
	while ((c = getopt_long(argc, argv, shortopts, longopts, NULL)) != -1) {
		err_exclusive_options(c, longopts, excl, excl_st);

		switch (c) {
		case 'c':
			opt_check = TRUE;
			break;
		case 'e':
			++n_opt_pipe;
			break;
		case 'f':
			++n_opt_pipe;
			break;
		case 'g':
			opt_get = TRUE;
			break;
		case 'i':
			++n_opt_pipe;
			break;
		case 'n':
			(void) strtos32_or_err(optarg, _("invalid fd argument"));
			++n_opt_pipe;
			break;
		case 'o':
			++n_opt_pipe;
			break;
		case 'q':
			opt_quiet = TRUE;
			break;
		case 's':
			sz = strtosize_or_err(optarg, _("invalid size argument"));
			opt_size = sz >= INT_MAX ? INT_MAX : (int)sz;
			++n_opt_size;
			break;
		case 'v':
			opt_verbose = TRUE;
			break;
		default:
			errtryhelp(EXIT_FAILURE);
		}
	}

	/* check arguments */
	if (opt_get) {
		if (argv[optind])
			errx(EXIT_FAILURE, _("cannot specify a command with --get"));

		/* print column headers, if requested */
		if (opt_verbose)
			printf("%s\t%s\t%s\n",
/* TRANSLATORS: a column that contains the names of files that are unix pipes */
				_("pipe"),
/* TRANSLATORS: a column that contains buffer sizes in bytes */
				_("size"),
/* TRANSLATORS: a column that contains an amount of data which has not been used by a program */
				_("unread")
			);

		/* special behavior for --get */
		if (!n_opt_pipe) {
			do_fd(STDIN_FILENO);
			return EXIT_SUCCESS;
		}
	} else {
		if (!set_size_default())
			goto execute_command;

		if (!opt_quiet && n_opt_size > 1)
			warnx(_("using last specified size"));

		/* special behavior for --set */
		if (!n_opt_pipe) {
			do_fd(STDOUT_FILENO);
			goto execute_command;
		}
	}

	/* go through the arguments again and do the requested operations */
	optind = 1;
	while ((c = getopt_long(argc, argv, shortopts, longopts, NULL)) != -1)
		switch (c) {
		case 'e':
			do_fd(STDERR_FILENO);
			break;
		case 'f':
			do_file(optarg);
			break;
		case 'i':
			do_fd(STDIN_FILENO);
			break;
		case 'n':
			/* optarg was checked before, but it's best to be safe */
			fd = strtos32_or_err(optarg, _("invalid fd argument"));
			do_fd(fd);
			break;
		case 'o':
			do_fd(STDOUT_FILENO);
			break;
		}

execute_command:
	/* exec the command, if it's present */
	if (!argv[optind])
		return EXIT_SUCCESS;

	execvp(argv[optind], &argv[optind]);
	errexec(argv[optind]);
}
